shadowsocks/udp.go

```diff
@@ -236,11 +236,30 @@ func (m *natmap) Add(clientAddr net.Addr, clientConn net.PacketConn, cipher shad
 	}()
 }
 
+// Get the length of the shadowsocks address header by parsing
+// addresses from the example range.
+var v6AddrLen int = len(socks.ParseAddr("[2001:db8::1]:12345"))
+var v4AddrLen int = len(socks.ParseAddr("192.0.2.1:12345"))
+
 // copy from src to dst at target with read timeout
 func timedCopy(clientAddr net.Addr, clientConn net.PacketConn, cipher shadowaead.Cipher, targetConn net.PacketConn,
 	timeout time.Duration, clientLocation, keyID string, sm metrics.ShadowsocksMetrics) {
-	textBuf := make([]byte, udpBufSize)
-	cipherBuf := make([]byte, udpBufSize)
+	// pkt is used for in-place encryption of downstream UDP packets, with the following layout.
+	// Receiving by proxy: | ------------------------------ | body | ---->
+	//
+	// If the server is IPv4, processing proceeds as follows:
+	// Prepared plaintext: | ---------- | server v4 address | body | ---->
+	// After encryption:   | --- | salt |        ciphertext        | tag | ---->
+	// Sent to client:           | salt |        ciphertext        | tag |
+	//
+	// If the server is IPv6, the address is longer, so the initial padding is not needed:
+	// Prepared plaintext: | ---- |    server v6 address    | body | ---->
+	// Sent to client:     | salt |              ciphertext        | tag |
+	pkt := make([]byte, udpBufSize)
+
+	saltSize := cipher.SaltSize()
+	// Leave enough room at the beginning of the packet for a max-length header (i.e. IPv6).
```

Comment:
Isn't the max length using domain names as the address? Is this assuming we will never have domain names downstream?

Reply:
Yes, that is the assumption.  The address we're writing here is the one returned by `ReadFrom`, which is the source address of a UDP packet, so it should never be a domain name.

If you want, we could make the type system enforce this by replacing all the uses of net.PacketConn with net.UDPConn, so we can use ReadFromUDP instead of ReadFrom.